### PR TITLE
Automatic pooling, refactor, and more

### LINKED ISSUE #9: Pools should handle connections to multiple hosts
The current implementation of connection pooling in `Mojito.Pool` is a naive Poolboy setup, with the limitation that the pool should only be used to connect to a single protocol+host+port.  Otherwise, the connection checked out of the pool is likely to be a connection to a different server, and we drop the existing connection and open another.

Consider a two-level pooling system, where a pool exists for each protocol+host+port we want to make requests to, and the process registry directs each request to its correct pool.

(Thanks to @ericmj for talking this through on Slack)

### PR DESCRIPTION
**Pull request by gamache**
This is a big PR.  It might be easier to review by just reading the whole codebase as it stands now, compared to reading a diff.

This is the release candidate for 0.3.0.  Closes #9.

* Lots of refactoring.
* `Mojito.request/1,5` is now the only request function presented to the user.  The `:pool` option controls whether automatic, manual, or no pooling is used.  Documentation doesn't mention manual pooling because users shouldn't; manual pooling may be removed entirely in the future.
* Pools are managed automatically by Mojito.  They are started when necessary, requests are matched to the correct pool, and more than one pool can be started for a given destination in order to reduce GenServer-related bottlenecks.
* I added a [benchmark](https://github.com/appcues/mojito/blob/autopool/BENCHMARK.md) because Mojito is fast as shit.

### CODE REVIEW

File: test/mojito/pool_test.exs [hunk 1]
```diff
@@ -1,82 +1,52 @@
...
+      tasks =
+        1..100
+        |> Enum.map(fn _ ->
+          Task.async(fn -> get("/wait?d=100", timeout: 500) end)
+        end)
 
-        ## 1 ready, 0 waiting, 1 in-progress (the one previously waiting)
-        assert({:ready, 1, 0, 1} = :poolboy.status(pool_name))
+      Task.yield_many(tasks, 600)
+      # |> IO.inspect()
```
Comment (andyleclair) [inline test/mojito/pool_test.exs]: A few stray `IO.inspect()` in here

Comment (gamache) [reply test/mojito/pool_test.exs]: Good eyes, thank you

File: lib/mojito/pool.ex [hunk 2]
```diff
@@ -1,151 +1,125 @@
...
+          | {:strategy, :lifo | :fifo}
+
+  @typep pool_key :: {String.t(), pos_integer}
+
+  @default_pool_opts [
+    size: 5,
+    max_overflow: 10,
+    pools: 5,
+    strategy: :lifo,
+    refractory_period: 3000
```
Comment (datbth) [inline lib/mojito/pool.ex]: This key seems wrong :thinking:

Comment (danhuynhdev) [inline lib/mojito/pool.ex]: This key seems wrong :thinking:

Comment (gamache) [reply lib/mojito/pool.ex]: Good eyes! That key did something (or tried to 😄) at some point in development.

**Pull request merged**